Updated to support Nest API changes and fix set_temp and away_mode an…

Zyell opened the pull request:
Hello.  I have made some changes to the Nest implementation in order to fix setting the temperature, grab the min and max temps from the Nest API, grab the correct Device name based on changes to the Nest API, and enforce sending all temperatures to the Nest API in celsius.  This requires the most recent version of python-nest, now at 2.4.0.  I have tested these changes with an actual Nest using the Home Assistant frontend interface.  Let me know if you would like me to make any changes to the request.  Thanks!

Zyell committed:
Updated to support Nest API changes and fix set_temp and away_mode and pull in new required python-nest version

balloob commented on homeassistant/components/thermostat/nest.py:
```diff
@@ -97,9 +106,16 @@ def is_away_mode_on(self):
     def set_temperature(self, temperature):
         """ Set new target temperature """
-        self.device.target = temperature
+        self.device.target = self.enforce_temp_units(temperature)
```
This should be solved in the component instead of the platform: we should only call `set_temperature` with the unit as specified in `unit_of_measurement`. I have extracted the helper methods that Home Assistant used internally so you can use them. They are available in the latest dev branch in the file [`homeassistant.helpers.temperature`](https://github.com/balloob/home-assistant/blob/086961d1093b466583c9c8125515981b242c0a2a/homeassistant/helpers/temperature.py) and [`homeassistant.util.temperature`](https://github.com/balloob/home-assistant/blob/086961d1093b466583c9c8125515981b242c0a2a/homeassistant/util/temperature.py)

balloob commented:
Awesome, thanks!

I've had one comment to be addressed.

You've also missed [one reference](https://github.com/Zyell/home-assistant/blob/master/homeassistant/components/thermostat/nest.py#L9) to the version of python-nest that is being used. Please update that one too.

Zyell committed:
Merge remote-tracking branch 'upstream/master'

Zyell committed:
updated nest to use helper method and updated requirements

Zyell commented:
Thanks for the comment, that makes a lot of more sense.  All changes have been made.  Let me know if there is anything else!

balloob commented on homeassistant/components/thermostat/nest.py:
```diff
@@ -98,7 +107,9 @@ def is_away_mode_on(self):
     def set_temperature(self, temperature):
-        """ Set new target temperature """
+        """ Set new target temperature ensuring it is in proper units """
+        temperature = convert(temperature, self.hass.config.temperature_unit,
```
I think that instead of calling `convert` here, we should move this call inside [this `for` loop](https://github.com/balloob/home-assistant/blob/master/homeassistant/components/thermostat/__init__.py#L88-L89):
```python
            for thermostat in target_thermostats:
                thermostat.set_temperature(convert(
                    temperature, self.hass.config.temperature_unit, thermostat.unit_of_measurement))
```

That way, it will be solved once and for all.

balloob commented on homeassistant/components/thermostat/nest.py:
```diff
@@ -109,6 +120,24 @@ def turn_away_mode_off(self):
+        """ Identifies mxn_temp in Nest API or defaults if not available. """
+        temp = self.device.away_temperature.high
+        if temp is None:
+            return super().max_temp
```
I think it should be that a thermostat should not be aware of the user `unit_of_measurement` preference. Instead it should always return all values in its own `unit_of_measurement`. It is the responsibility of the class `Thermostat` to convert it when it is exposing it to the outside world.

This means that the conversion from thermostat to state has to happen [over here](https://github.com/balloob/home-assistant/blob/master/homeassistant/components/thermostat/__init__.py#L136-L137).

We can then convert the `min_temp` default value in the `Thermostat` class to be:
```python
    @property
    def min_temp(self):
        """ Return minimum temperature. """
        return convert(7, TEMP_CELCIUS, self.unit_of_measurement)
```

To make sure that every `Thermostat` object has a `unit_of_measurement`, we should probably add it to the `Thermostat` class:
```python
    @property
    def unit_of_measurement(self):
        """ Unit of measurement this thermostat expresses itself in. """
        raise NotImplementedError
```

Zyell committed:
Have thermostat class make all necessary unit conversions

Zyell commented:
Ok.  I think this is much better.  I have moved all unit conversions to the Thermostat class and cleaned up the the state_attributes method a little.

balloob commented:
Awesome!

balloob commented:
Oh snap, just realised your PR points at the master branch. Can you point it at the dev branch? For this you have to close it and open a new one.

Zyell commented:
Sorry about that.  Still getting used to using git and github...
